### ISSUE TITLE
True random access (i.e. Dataset vs. IterableDataset)

### ISSUE THREAD

**Issue by jotaf98** [tmbdev/webdataset#48]
Hi, this is a really neat library! However it would be nice to have a simpler interface to TAR files that allows random access, instead of enforcing sequential access. Let me explain why.

The most common use-case for academic labs such as mine are not terabyte-scale remote datasets where sharding is common, but several gigabytes datasets with a few million small images.

So, a dataset fits in a single machine; but due to automatic scheduling in a cluster we cannot have all datasets in all machines. In this context, the easiest solution is to copy the dataset to the local machine at the start of training. However, millions of small files really strain the filesystem (both for copying and accessing during training). So copying and reading from a single TAR file would be ideal -- and WebDataset seems (on the surface) to do this well.

But the constraints of the IterableDataset are maybe a step too far. We now have to make decisions about how many shards to use, sizes of rolling buffers for shuffling, etc. This adds a ton of friction, and the uneasy feeling that now the samples are not sufficiently IID for training if you make a wrong decision. Compare this to the ease of training with tons of small images in a filesystem.

I was trying to use WebDataset and get colleagues to adopt it, but this is a big wall for adoption. Uncompressed TAR files allow random access. Could this not be a simpler entry-point for WebDataset? A user who wants to scale things up would find it easier to adapt then to the IterableDataset, but I think that many users would be perfectly happy with the random access version, which is much less restrictive.

**Comment by tmbdev** [comment]
I agree that it might make users feel better initially if there was a bridge. The only real difference I see is that random access has the appearance of a better shuffle (in practice, there is no difference) and that people don't have to shard, but at the cost of substantially slower I/O.

If your data is already stored in a single .tar file, sharding is trivial using the companion tarp tools (github.com/tmbdev/tarp):

tarp split -c 1000 onefile.tar -

I'll keep it open as an RFE, but I don't know when it will happen.

**Comment by jotaf98** [comment]
Thanks! To make sure I understood your comments on performance, we are comparing 3 scenarios:

1. Random access to many small files, the current PyTorch default (which most people use).
2. Random access to a single TAR file, suggested here.
3. Sequential access to a few TAR files, implemented in WebDataset.

All with worker threads.

I expect the performance ordering to be 1 < 2 < 3 but it's a question of by how much. For users of 1, moving to 2 would already be an improvement on faster dataset transfers, lower inode use, etc, but loading from a local disk might still incur a penalty due to seek times, so overall we wouldn't expect it to be substantially faster. Option 3 would, but it involves more changes to workflow and storage.

So while option 2 could be "substantially slower" than option 3, it's not any worse than option 1 which is what they use now. I guess users who are not looking to squeeze that last bit of performance will probably be happy with the plug-and-play nature of option 2.

**Comment by tmbdev** [comment]
I'm still not really sure, though, what you are looking for. You can certainly write an indexable Dataset class that accesses a .tar file containing training data using random access tar libraries. That would simply end up being a separate library, a kind of companion to WebDataset, since it would share very little code with WebDataset (which is built around stream abstractions).

But a simple way of dealing with moderate numbers of workers on a single machine is something like:

```
def nosplit(l): return l
worker_id, num_workers = ... torch.utils.data.get_worker_info() ...
dataset = WebDataset(my_tar_file, splitter=nosplit).slice(worker_id, 999999999, num_workers).shuffle(10000).decode("rgb")
```

That way, you can use a single .tar file and each worker sees a distinct subset of the .tar file. Up to about 10 workers, it's probably as efficient as random access.

**Comment by jotaf98** [comment]
Thanks, that's great! I'll have to run some speed comparisons of all the different methods when I get some time.

You have a really good point that, since WebDataset is built around streams, anything of this sort would actually not reuse much code.

Do you think that a PR with a simple implementation of it (like [this](https://discuss.pytorch.org/t/how-to-read-dataset-from-tar-files/47425/2)) would make sense for WebDataset?

On one hand, the goals of both schemes are pretty much aligned. On the other hand, the technical solutions to achieve them are (IterableDataset vs indexable Dataset) are orthogonal.
